File: 064_minimum_path_sum/Solution1/Solution.java
class Solution {
    public int minPathSum(int[][] grid) {
        if (grid == null || grid.length == 0 || grid[0] == null || grid[0].length == 0)
            return 0;

        int m = grid.length, n = grid[0].length;
        int[][] dp = new int[m][n];

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (i == 0 && j == 0) dp[0][0] = grid[0][0];
                else if (i == 0) dp[0][j] = dp[0][j - 1] + grid[0][j];
                else if (j == 0) dp[i][0] = dp[i - 1][0] + grid[i][0];
                else dp[i][j] = grid[i][j] + Math.min(dp[i - 1][j], dp[i][j - 1]);
            }
        }

        return dp[m - 1][n - 1];
    }

    public static void main(String[] args) {
        int[][] grid = {
            {1, 3, 1},
            {1, 5, 1},
            {4, 2, 1}
        };
        System.out.println(new Solution().minPathSum(grid));
    }
}

File: 064_minimum_path_sum/Solution1/Solution.py
class Solution:
    def minPathSum(self, grid):
        if not grid or len(grid) == 0 or not grid[0] or len(grid[0]) == 0:
            return 0

        m, n = len(grid), len(grid[0])
        dp = [[0] * n for _ in range(m)]

        for i in range(m):
            for j in range(n):
                if i == 0 and j == 0: dp[0][0] = grid[0][0]
                elif i == 0: dp[0][j] = dp[0][j - 1] + grid[0][j]
                elif j == 0: dp[i][0] = dp[i - 1][0] + grid[i][0]
                else: dp[i][j] = grid[i][j] + min(dp[i - 1][j], dp[i][j - 1])

        return dp[-1][-1]


if __name__ == '__main__':
    grid = [
        [1, 3, 1],
        [1, 5, 1],
        [4, 2, 1]
    ]
    print(Solution().minPathSum(grid))

